proxy.go

```diff
@@ -0,0 +1,292 @@
+
+	var port = msg.LocalPort
+	if port == 0 {
+		log.Warning("TCP proxy to port 0")
+	}
+
+	var localAddr = fmt.Sprintf("127.0.0.1:%d", port)
+	if p.LocalAddr != "" {
+		localAddr = p.LocalAddr
+	} else if p.FetchLocalAddr != nil {
+		l, err := p.FetchLocalAddr(msg.LocalPort)
+		if err != nil {
+			log.Warning("Failed to get custom local address: %s", err)
+			return
+		}
+		localAddr = l
+	}
+
+	log.Debug("Dialing local server: %q", localAddr)
+	local, err := net.DialTimeout("tcp", localAddr, defaultTimeout)
+	if err != nil {
+		log.Error("Dialing local server %q failed: %s", localAddr, err)
+		return
+	}
+
+	Join(local, remote, log)
+}
+
+func (p *ProxyTCP) log() logging.Logger {
+	if p.Log != nil {
+		return p.Log
+	}
+	return tpcLog
+}
+
+// Join copies data between local and remote connections.
+// It reads one connection and writes to the other.
+// It aims to provide a building block for custom Proxy implementations.
+func Join(local, remote net.Conn, log logging.Logger) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	transfer := func(side string, dst, src net.Conn) {
+		log.Debug("proxing %s -> %s", src.RemoteAddr(), dst.RemoteAddr())
+
+		n, err := io.Copy(dst, src)
+		if err != nil {
+			log.Error("%s: copy error: %s", side, err)
+		}
+
+		if err := src.Close(); err != nil {
+			log.Debug("%s: close error: %s", side, err)
+		}
+
+		// not for yamux streams, but for client to local server connections
+		if d, ok := dst.(*net.TCPConn); ok {
+			if err := d.CloseWrite(); err != nil {
+				log.Debug("%s: closeWrite error: %s", side, err)
+			}
+		}
+
+		wg.Done()
+		log.Debug("done proxing %s -> %s: %d bytes", src.RemoteAddr(), dst.RemoteAddr(), n)
+	}
+
+	go transfer("remote to local", local, remote)
+	go transfer("local to remote", remote, local)
+
+	wg.Wait()
+}
+
+////////////////////
+// ProxyOverwrite //
+////////////////////
+
+// ProxyOverwrite enables easy setting of different proxy functions for different
+// transport protocols. Consider the following example:
+//
+//     tunnel.ProxyOverwrite{
+//             HTTP: &tunnel.ProxyHTTP{
+//                     LocalAddr: localAddr,
+//             },
+//             WS: &MyCustomWSProxy{},
+//     }
+//
+// This code would result in a Proxy implementation that:
+//
+// * forwards all HTTP calls to localAddr
+// * uses MyCustomWSProxy for web sockets
+// * handles TCP using default implementation
+type ProxyOverwrite struct {
+	// HTTP is optional custom implementation of HTTP proxing.
+	HTTP Proxy
+	// TCP is optional custom implementation of TCP proxing.
+	TCP Proxy
+	// WS is optional custom implementation of web socket proxing.
+	WS Proxy
+
+	defaultHTTP ProxyHTTP
+	defaultTCP  ProxyTCP
+}
+
+// Proxy selects appropriate Proxy method based on control message protocol.
+// It proxy method for a given protocol was not ovewritten a default implementation
+// would be used.
+func (p *ProxyOverwrite) Proxy(remote net.Conn, msg *proto.ControlMessage) {
+	switch msg.Protocol {
```

Comment:
Most likely we want to close the net.Conn here id default clause if we can't proxy it.